controller/controller.go

```diff
@@ -208,15 +208,29 @@ type Impl struct {
 // NewImpl instantiates an instance of our controller that will feed work to the
 // provided Reconciler as it is enqueued.
 func NewImpl(r Reconciler, logger *zap.SugaredLogger, workQueueName string) *Impl {
```

Comment:
This is starting to look like a good candidate for [the options pattern](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis).

Reply:
+1, I can switch.

Reply:
They should share the RL I think? So just one of them?

Reply:
Not sure I understand your comment.

My issues is that there is no func(blah *blah) that can modify the ratelimiter, as it's private on the rate limiting queue interface.

So really the only 'options' pattern I can think of is

1. a struct with the ratelimiter on it
2. an interface that can return a ratelimiter

and I don't see any reason to prefer those over this, unless y'all agree 1 is better (with an Options that can contain a statslogger and/or ratelimiter)

Reply:
Not sure I follow, just NewImpl(opts Options).
We should not modify rate limiter. We should either pass one created from outside, or just use the default (what we do right now).

Reply:
Added an options at the end, wdyt?

Anything further and we risk signficant backward incompatible changes, which I don't want to take the responsibility for rolling out across the federated repos.

Reply:
Ah, Lol, NewImplFull is one I made up. Sure, Options can include the other stuff

Reply:
I was trying to suggest

```go
type Option func(*Impl)

func NewImpl(r Reconciler, logger *zap.SugaredLogger, workQueueName string, opts ...Option) {
  ret := &Impl{
    Name: workQueueName,
    Reconciler: r,
    logger: logger.Named(workQueueName),
    statsReporter: MustNewStatsReporter(workQueueName, logger),  // assumes statsreporter is cheap to create
  }
  for _, o := range opts {
    o.Apply(ret)
  }
  if ret.workQueue == nil {  // Avoid creating a new workqueue if that is expensive
    ret.workQueue: newTwoLaneWorkQueue(workQueueName, workqueue.DefaultControllerRateLimiter())
  }
  return ret
}

func WithLimit(rateLimiter *workqueue.RateLimiter) Option {
  return func(i *Impl) {
    i.workQueue = newTwoLaneWorkQueue(i.Name, rateLimiter)
  }
}

func WithReporter(reporter StatsReporter) Option {
  return func(i *Impl) {
    i.statsReporter = reporter
  }
}
```

But I think this form is fine, too.